testing/test_service_assertions.py

```diff
@@ -0,0 +1,231 @@
+
+
+class BaseService(ABC):
+    @abstractmethod
+    def service_method(self, data):
+        pass
+
+
+class SomeService(BaseService):
+    def service_method(self, data):
+        self.data = data
+
+
+# class SomeStepThatChangesTheRootOfTheExecutionContext(ForceHandleMixin, Identity):
+#     def __init__(self):
+#         Identity.__init__(self)
+#         ForceHandleMixin.__init__(self)
+#
+#     def _will_process(self, data_container: DataContainer, context: ExecutionContext):
+#         context.root = 'invalid_root'
+#         super()._will_process(data_container, context)
+#         return data_container, context
+#
+#     def _transform_data_container(self, data_container: DataContainer, context: ExecutionContext):
+#         return data_container
+
+
+class SomeStep(ForceHandleMixin, Identity):
+    def __init__(self):
+        Identity.__init__(self)
+        ForceHandleMixin.__init__(self)
+
+    def _will_process(self, data_container: DataContainer, context: ExecutionContext):
+        data_container, context = super()._will_process(data_container, context)
+        service = context.get_service(BaseService)
+        return data_container, context
+
+    def _fit_data_container(self, data_container: DataContainer, context: ExecutionContext):
+        service: BaseService = context.get_service(BaseService)
+        service.service_method(data_container.data_inputs)
+        return self
+
+    def _transform_data_container(self, data_container: DataContainer, context: ExecutionContext):
+        service: BaseService = context.get_service(BaseService)
+        service.service_method(data_container.data_inputs)
+        return data_container
+
+
+class RegisterServiceDynamically(ForceHandleMixin, Identity):
+    def __init__(self):
+        Identity.__init__(self)
+        ForceHandleMixin.__init__(self)
+
+    def _will_process(self, data_container: DataContainer, context: ExecutionContext):
+        # data_container, context = ForceHandleMixin._will_process(data_container, context)
+        context.register_service(BaseService, SomeService())
+        return data_container, context
+
+    def _transform_data_container(self, data_container: DataContainer, context: ExecutionContext):
+        return data_container
+
+
+def test_step_with_context_should_only_save_wrapped_step(tmpdir):
+    context = ExecutionContext(root=tmpdir)
+    service = SomeService()
+    context.set_service_locator({BaseService: service})
+    p = Pipeline([
+        SomeStep().assert_has_services(BaseService)
+    ]).with_context(context=context)
+
+    p.save(context, full_dump=True)
+
+    p: Pipeline = ExecutionContext(root=tmpdir).load(os.path.join('StepWithContext', 'Pipeline'))
+    assert isinstance(p, Pipeline)
+
+
+def test_with_context_should_inject_dependencies_properly(tmpdir):
+    data_inputs = np.array([0, 1, 2, 3])
+    context = ExecutionContext(root=tmpdir)
+    service = SomeService()
+    context.set_service_locator({BaseService: service})
```

Comment:
```suggestion
    context.set_service_locator({SomeBaseService: service})
```

Reply:
I disagree with this change. BaseService is the abstract class and SomeBaseService its implementation. Conceptually, its better to use the abstract class as a key. 

Plus, if we'd do that change, we'd have to change BaseService for SomeBaseService in every get_service and assert_has_service call.

Reply:
In my suggestion, the class SomeService would still inherit from SomeBaseService. My fear was that the class BaseService be confusing as wheter every neuraxle service on earth would need to inherit from this class if it was ever to be moved from the test file to the base.py file for instance. Anyway the BaseService is strictly limited to the test file's scope here and shouldn't be moved elsewhere quickly. 

In otherwords, my point was that even abstract classes, when they are for testing, should make the usage of the keyword "Some" as they are only some examples of abstract classes. So after applying my comment we'd have `class SomeService(SomeBaseService)` and still keep some class hierarchy with some abstract class.

Reply:
Make sense. I'll name it SomeBaseService